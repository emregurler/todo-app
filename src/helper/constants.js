const filterTypeMap = {
  ALL: 'all',
  DONE: 'done',
  ACTIVE: 'active',
};

const templateModeMap = {
  EDIT: 'edit',
  ADD: 'add',
  LIST: 'list',
};

const taskStatusMap = {
  FAIL: 'fail',
  SUCCESS: 'success',
  NEUTRAL: 'neutral',
  HURRY: 'hurry',
};

const deadlineTextMap = {
  fail: 'You are late :(',
  success: 'Congrats!',
  neutral: '',
  hurry: "Hurry up, it's almost time!",
};

export { filterTypeMap, templateModeMap, taskStatusMap, deadlineTextMap };
