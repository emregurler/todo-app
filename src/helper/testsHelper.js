import { reducerInitialState } from '../store/services/todo/reducer';

export const todosLenght_1 = [
  {
    id: 1,
    title: 'Todo1',
    decription: 'Description1',
    deadline: '2021-07-04 22:00:00',
    done: false,
  },
];

export const todosLenght_2 = [
  {
    id: 1,
    title: 'Todo1',
    decription: 'Description1',
    deadline: '2021-07-04 22:00:00',
    done: false,
  },
  {
    id: 2,
    title: 'Todo2',
    decription: 'Description2',
    deadline: '2021-07-04 22:00:00',
    done: false,
  },
];

export const todoReducer_2 = {
  ...reducerInitialState,
  allTodos: todosLenght_2,
  hasData: true,
};
