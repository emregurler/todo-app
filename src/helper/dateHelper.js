import dayjs from 'dayjs';
import { taskStatusMap } from './constants';

const defaultDateFormat = 'YYYY-MM-DD HH:mm:ss';

const getTaskStatusWithDate = (taskDate, isDone, dateFormat = defaultDateFormat) => {
  const now = dayjs();
  const checkingDate = dayjs(taskDate, dateFormat);

  const difference = checkingDate.diff(now);
  if (difference > 0) {
    if (isDone) {
      return taskStatusMap.SUCCESS;
    } else {
      const nowPlusThreeHour = now.add(3, 'hour');
      const isInLastThreeHour = checkingDate.diff(nowPlusThreeHour) <= 0;
      return isInLastThreeHour ? taskStatusMap.HURRY : taskStatusMap.NEUTRAL;
    }
  } else {
    return taskStatusMap.FAIL;
  }
};

export { getTaskStatusWithDate, defaultDateFormat };
