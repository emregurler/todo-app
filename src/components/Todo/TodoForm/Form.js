import React, { useEffect } from 'react';
import { Form as AntForm, Input, Button, DatePicker, Row, Space, Divider } from 'antd';
import { ArrowLeftOutlined } from '@ant-design/icons';
import moment from 'moment';
import PropTypes from 'prop-types';
import { defaultDateFormat } from '../../../helper/dateHelper';

const { TextArea } = Input;

const Form = ({ todo, onSubmit, onBack }) => {
  const [form] = AntForm.useForm();

  useEffect(() => {
    if (todo) {
      const deadline = moment(todo.deadline, defaultDateFormat);
      form.setFieldsValue({ ...todo, deadline });
    } else {
      handleReset();
    }
  }, [JSON.stringify(todo)]);

  const tailLayout = {
    wrapperCol: { offset: 6, span: 16 },
  };

  const handleReset = () => {
    form.resetFields();
  };

  return (
    <AntForm
      layout="vertical"
      onFinish={onSubmit}
      form={form}
      initialValues={{ layout: 'horizontal' }}
    >
      <AntForm.Item
        name="title"
        label="Title"
        rules={[{ required: true, message: "Please enter todo's title!", whitespace: true }]}
      >
        <Input placeholder="Enter title" />
      </AntForm.Item>
      <AntForm.Item name="description" label="Description">
        <TextArea rows={4} placeholder="Enter description" />
      </AntForm.Item>
      <AntForm.Item
        name="deadline"
        label="Deadline"
        rules={[{ type: 'object', required: true, message: 'Please select time!' }]}
      >
        <DatePicker format="YYYY-MM-DD HH:mm:ss" showTime />
      </AntForm.Item>
      <Divider style={{ border: 'none' }} />
      <AntForm.Item>
        <Row justify="space-between">
          <Button
            icon={<ArrowLeftOutlined />}
            size="large"
            htmlType="button"
            onClick={() => {
              onBack(() => {
                handleReset();
              });
            }}
          >
            Back
          </Button>

          <Space>
            <Button style={{ padding: '0 40px' }} size="large" type="primary" htmlType="submit">
              {todo ? 'Update' : 'Add'}
            </Button>
            <Button size="large" htmlType="button" onClick={handleReset}>
              Reset Fields
            </Button>
          </Space>
        </Row>
      </AntForm.Item>
    </AntForm>
  );
};

Form.propTypes = {
  todo: PropTypes.object,
  onSubmit: PropTypes.func,
  onBack: PropTypes.func,
};

export default Form;
