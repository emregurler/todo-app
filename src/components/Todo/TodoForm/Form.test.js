import React from 'react';
import { render } from '@testing-library/react';
import Form from '../TodoForm/Form';
import { todosLenght_1 } from '../../../helper/testsHelper';

describe('Form component tests starting...', () => {
  it('form should displayed empty if it is add mode', () => {
    const formComponent = () => render(<Form />);
    expect(formComponent).toMatchSnapshot();
  });

  it('form should displayed empty if it is add mode', () => {
    const todo = { ...todosLenght_1[0] };
    const formComponent = () => render(<Form todo={todo} />);
    expect(formComponent).toMatchSnapshot();
  });
});

// it('template should have class if has prop', () => {
//     const templateComponent = () => render(<Template className="newClass" />);
//     const { getByTestId } = templateComponent();
//     const template = getByTestId('template');
//     expect(template.classList.contains('newClass')).toBe(true);
//   });

//   it('template should only have container class if className not given', () => {
//     const templateComponent = () => render(<Template />);
//     const { getByTestId } = templateComponent();
//     const template = getByTestId('template');
//     expect(Array.from(template.classList).every((cl) => cl === 'container')).toBe(true);
//   });

//   it('template header should have title if given', () => {
//     const templateHeaderComponent = () => render(<TemplateHeader title="Todo App" />);
//     const { getByTestId } = templateHeaderComponent();
//     const templateHeader = getByTestId('template-title');
//     expect(templateHeader).toHaveTextContent('Todo App');
//   });
