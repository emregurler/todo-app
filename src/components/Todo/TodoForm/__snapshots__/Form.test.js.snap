// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component tests starting... form should displayed empty if it is add mode 1`] = `[Function]`;

exports[`Form component tests starting... form should displayed empty if it is add mode 2`] = `[Function]`;
